src/gui/TorchMenu.cxx

```diff
@@ -0,0 +1,271 @@
+            main_gui->SetVisible(!main_gui->IsVisible());
+            input->SetMouseMode(main_gui->IsVisible() ? Urho3D::MM_ABSOLUTE : Urho3D::MM_RELATIVE);
+        }
+    } else {
+        send_compile_event();
+    }
+}
+
+void TorchMenu::handle_ino_set(Urho3D::StringHash, Urho3D::VariantMap&) {
+    ino_path = get_selector_path();
+    update_runtime_values();
+    selector.Reset();
+    send_compile_event();
+}
+
+void TorchMenu::handle_smce_home_set(Urho3D::StringHash, Urho3D::VariantMap&) {
+    options.smce_home = get_selector_path().parent_path();
+    update_runtime_values();
+    selector.Reset();
+}
+
+void TorchMenu::handle_veh_conf_set(Urho3D::StringHash, Urho3D::VariantMap&) {
+    options.vehicle_config = get_selector_path();
+    update_runtime_values();
+    selector.Reset();
+}
+
+void TorchMenu::handle_board_conf_set(Urho3D::StringHash, Urho3D::VariantMap&) {
+    options.board_config = get_selector_path();
+    update_runtime_values();
+    selector.Reset();
+}
+
+void TorchMenu::handle_popup_ok(Urho3D::StringHash, Urho3D::VariantMap&) {
+    if (popup)
+        popup->Remove();
+    popup.Reset();
+}
+
+void TorchMenu::make_popup(std::string warn_title, std::string warning) {
+    popup = ui->LoadLayout(cache->GetResource<Urho3D::XMLFile>("Torch/Data/UI/TorchPopup.xml"), styling);
+    popup->SetPriority(10000);
+    popup->GetChildDynamicCast<Urho3D::Text>("popup_title", true)->SetText(warn_title.c_str());
+    popup->GetChildDynamicCast<Urho3D::Text>("popup_text", true)->SetText(warning.c_str());
+    auto* hello = popup->GetChild("ok_button", true);
+    SubscribeToEvent(hello, Urho3D::E_CLICKEND, URHO3D_HANDLER(TorchMenu, handle_popup_ok));
+    main_gui->AddChild(popup);
+}
+
+void TorchMenu::send_compile_event() {
+    if (ino_path && std::filesystem::exists(options.vehicle_config) && std::filesystem::exists(options.board_config) &&
+        emu_state == EmuStates::NO_SKETCH) {
+        auto* load_button = main_gui->GetChild("load_button", true);
+        load_button->SetEnabled(false);
+        main_gui->GetChildDynamicCast<Urho3D::Text>("load_button_text", true)->SetText("Compiling..");
+
+        Urho3D::VariantMap map;
+        // VariantMap requires RefCounted so convert to Urho3D::String
+        map[sketchCompileData::P_INO_PATH] = Urho3D::String(ino_path->c_str());
+        map[sketchCompileData::P_VEH_CONFIG_PATH] = Urho3D::String(options.vehicle_config.c_str());
+        map[sketchCompileData::P_BOARD_CONFIG_PATH] = Urho3D::String(options.board_config.c_str());
+        map[sketchCompileData::P_SMCE_HOME_PATH] = Urho3D::String(options.smce_home.c_str());
+
+        SendEvent(E_SKETCH_COMPILE_DATA, map);
+    } else {
+        make_popup("Compile Error", "Cannot compile without a board and vehicle config.");
+    }
+}
+void TorchMenu::handle_console_redirect(Urho3D::StringHash, Urho3D::VariantMap&) {
+    if (auto* line_edit = console->GetLineEdit(); line_edit) {
```

Comment:
Pretty sure this can be reduced to `if (auto* line_edit = console->GetLineEdit())`